consensus/tendermint/tendermint.go

```diff
@@ -0,0 +1,436 @@
+					return
+				case t.prevotesCh <- vote:
+				}
+			}
+		}
+	}
+
+	if len(precommits) > 0 {
+		for _, addrPrecommits := range precommits {
+			for _, vote := range addrPrecommits {
+				select {
+				case <-t.quit:
+					return
+				case t.precommitsCh <- vote:
+				}
+			}
+		}
+	}
+
+	t.futureMessages.deleteRoundMessages(h, r)
+}
+
+type timeout struct {
+	*time.Timer
+
+	s step
+	h uint
+	r uint
+}
+
+func (t *Tendermint[V, H, A]) scheduleTimeout(duration time.Duration, s step, h, r uint) {
+	tm := timeout{s: s, h: h, r: r}
+	tm.Timer = time.AfterFunc(duration, func() {
+		select {
+		case <-t.quit:
+		case t.timeoutsCh <- tm:
+		}
+	})
+	t.scheduledTms = append(t.scheduledTms, tm)
+}
+
+func (t *Tendermint[_, H, A]) OnTimeoutPropose(h, r uint) {
+	if t.state.height == h && t.state.round == r && t.state.step == propose {
+		vote := Prevote[H, A]{
+			Vote: Vote[H, A]{
+				Height: t.state.height,
+				Round:  t.state.round,
+				ID:     nil,
+				Sender: t.nodeAddr,
+			},
+		}
+		t.messages.addPrevote(vote)
+		t.broadcasters.PrevoteBroadcaster.Broadcast(vote)
+		t.state.step = prevote
+	}
+}
+
+func (t *Tendermint[_, H, A]) OnTimeoutPrevote(h, r uint) {
+	if t.state.height == h && t.state.round == r && t.state.step == prevote {
+		vote := Precommit[H, A]{
+			Vote: Vote[H, A]{
+				Height: t.state.height,
+				Round:  t.state.round,
+				ID:     nil,
+				Sender: t.nodeAddr,
+			},
+		}
+		t.messages.addPrecommit(vote)
+		t.broadcasters.PrecommitBroadcaster.Broadcast(vote)
+		t.state.step = precommit
+	}
+}
+
+func (t *Tendermint[_, _, _]) OnTimeoutPrecommit(h, r uint) {
+	if t.state.height == h && t.state.round == r {
+		t.startRound(r + 1)
+	}
+}
+
+// line55 assumes the caller has acquired a mutex for accessing future messages.
+/*
+	55: upon f + 1 {∗, h_p, round, ∗, ∗} with round > round_p do
+	56: 	StartRound(round)
+*/
+func (t *Tendermint[V, H, A]) line55(futureR uint) {
+	vals := make(map[A]struct{})
+	proposals, prevotes, precommits := t.futureMessages.allMessages(t.state.height, futureR)
+
+	// If a validator has sent proposl, prevote and precommit from a future round then it will only be counted once.
+	for addr := range proposals {
+		vals[addr] = struct{}{}
+	}
+
+	for addr := range prevotes {
+		vals[addr] = struct{}{}
+	}
+
+	for addr := range precommits {
+		vals[addr] = struct{}{}
+	}
+
+	if t.validatorSetVotingPower(slices.Collect(maps.Keys(vals))) > f(t.validators.TotalVotingPower(t.state.height)) {
+		t.startRound(futureR)
+	}
+}
+
+func (t *Tendermint[V, H, A]) validatorSetVotingPower(vals []A) uint {
+	var totalVotingPower uint
+	for _, v := range vals {
+		totalVotingPower += t.validators.ValidatorVotingPower(v)
+	}
+	return totalVotingPower
+}
+
+// Todo: add separate unit tests to check f and q thresholds.
+func f(totalVotingPower uint) uint {
+	// note: integer division automatically floors the result as it return the quotient.
+	return (totalVotingPower - 1) / 3
+}
+
+func q(totalVotingPower uint) uint {
```

Comment:
Shouldn't it be `2f + 1`?

Reply:
That is not exactly the case. Please read the Threshold section of Juno's Tendermint implementation.

Reply:
I'd proposed using a simplified version of this:
```
	return (totalVotingPower + f(totalVotingPower)) / 2 + 1
```

Reply:
This is based on the original assumption 2Q > N +f